File: lintcode/532_reverse-pairs/reverse-pairs.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   chenweikang
@Problem:  http://www.lintcode.com/problem/reverse-pairs
@Language: Python
@Datetime: 16-05-03 00:03
'''

class Solution:
    # @param {int[]} A an array
    # @return {int} total of reverse pairs
    def reversePairs(self, A):
        def merge(arr_l, arr_r):
            m, n = len(arr_l), len(arr_r)
            i = j = 0
            arr_a = []
            rev_m = 0
            while i < m and j < n:
                if arr_l[i] <= arr_r[j]:
                    arr_a.append(arr_l[i])
                    i += 1
                else:
                    arr_a.append(arr_r[j])
                    j += 1
                    rev_m += m - i
            while i < m:
                arr_a.append(arr_l[i])
                i += 1
            while j < n:
                arr_a.append(arr_r[j])
                j += 1
            return arr_a, rev_m

        def util(A):
            n = len(A)
            if n <= 1:
                return A, 0
            m = n / 2;
            arr_l, rev_l = util(A[:m])
            arr_r, rev_r = util(A[m:])
            arr_a, rev_m = merge(arr_l, arr_r)
            return arr_a, rev_l + rev_r + rev_m
        
        sortedA, res = util(A)
        return res

File: lintcode/532_reverse-pairs/reverse-pairs.cpp
/*
@Copyright:LintCode
@Author:   chenweikang
@Problem:  http://www.lintcode.com/problem/reverse-pairs
@Language: C++
@Datetime: 16-05-19 14:24
*/

class Solution {
public:
    /**
     * @param A an array
     * @return total of reverse pairs
     */
    long long reversePairs(vector<int>& A) {
        long long rev_pairs = 0;
        int n = A.size();
        if (n < 2) return 0;
        mergesort(A, 0, n - 1, rev_pairs);
        return rev_pairs;
    }

    void mergesort(vector<int>& A, int l, int r, long long &rev_pairs) {
        if (l >= r) return;
        int m = l + (r - l) / 2;
        mergesort(A, l, m, rev_pairs);
        mergesort(A, m+1, r, rev_pairs);
        merge(A, l, m, r, rev_pairs);
    }

    void merge(vector<int>& A, int l, int m, int r, long long &rev_pairs) {
        vector<int> vl(A.begin() + l, A.begin() + m + 1);
        vector<int> vr(A.begin() + m + 1, A.begin() + r + 1);
        const int sl = m - l + 1;
        const int sr = r - m;
        int i = 0, j = 0, k = l;
        while (i < sl && j < sr) {
            if (vl[i] <= vr[j]) {
                A[k++] = vl[i++];
            } else {
                A[k++] = vr[j++];
                rev_pairs += sl - i;
            }
        }
        while (i < sl)
            A[k++] = vl[i++];
        while (j < sr)
            A[k++] = vr[j++];
    }
};